src/java.base/share/classes/jdk/internal/util/ArraysSupport.java

```diff
@@ -575,52 +575,84 @@ public static int mismatch(double[] a, int aFromIndex,
+     * amount, and a preferred growth amount. The computation is done in an overflow-safe
+     * fashion.
+     *
+     * This method is used by objects that contain an array that might need to be grown
+     * in order to fulfill some immediate need (the minimum growth amount) but would also
+     * like to request more space (the preferred growth amount) in order to accommodate
+     * potential future needs. The returned length is usually clamped at the soft maximum
+     * length in order to avoid hitting the JVM implementation limit. However, the soft
+     * maximum will be exceeded if the minimum growth amount requires it.
+     *
+     * If the preferred growth amount is less than the minimum growth amount, the
+     * minimum growth amount is used as the preferred growth amount.
+     *
+     * The preferred length is determined by adding the preferred growth amount to the
+     * current length. If the preferred length does not exceed the soft maximum length
+     * (SOFT_MAX_ARRAY_LENGTH) then the preferred length is returned.
      *
-     * @param oldLength   current length of the array (must be non negative)
-     * @param minGrowth   minimum required growth of the array length (must be
-     *                    positive)
-     * @param prefGrowth  preferred growth of the array length (ignored, if less
-     *                    then {@code minGrowth})
+     * Since the preferred length exceeds the soft maximum, we use the minimum growth
+     * amount. The minimum required length is determined by adding the minimum growth
+     * amount to the current length. If the minimum required length is less than the
+     * soft maximum, the soft maximum is returned. If the minimum required length is
+     * greater than the soft maximum but does not exceed Integer.MAX_VALUE, the minimum
+     * required length is returned. Otherwise, the minimum required length exceeds
+     * Integer.MAX_VALUE, which can never be fulfilled, so this method throws OutOfMemoryError.
+     *
+     * Note that this method does not do any array allocation itself; it only does array
+     * length growth computations. However, it will throw OutOfMemoryError as noted above.
+     *
+     * Note also that this method cannot detect the JVM's implementation limit, and it
+     * may compute and return a length value up to and including Integer.MAX_VALUE that
+     * might exceed the JVM's implementation limit. In that case, the caller will likely
+     * attempt an array allocation with that length and encounter an OutOfMemoryError.
+     * Of course, regardless of the length value returned from this method, the caller
+     * may encounter OutOfMemoryError if there is insufficient heap to fulfill the request.
+     *
+     * @param oldLength   current length of the array (must be nonnegative)
+     * @param minGrowth   minimum required growth amount (must be positive)
+     * @param prefGrowth  preferred growth amount
      * @return the new length of the array
```

Comment:
Slightly better seems s/@return the new length of the array/@return the new array length/